### PR TITLE
ExoCom: Show response time of service

### PR DESCRIPTION
**Pull request by martinjaime**
@kevgo @helenmedrano 

I had not noticed I had been unassigned, but I just finished this issue. This pull request is paired with [/Originate/exosphere-sdk#139](https://github.com/Originate/exosphere-sdk/pull/139)

Addresses [/Originate/exosphere-sdk#30](https://github.com/Originate/exosphere-sdk/issues/30)

### CODE REVIEW

File: features/broadcasting-messages.feature [hunk 1]
```diff
@@ -25,5 +25,5 @@ Feature: Broadcasting messages
 
   Scenario: broadcasting a reply
     When the "users" service sends "created-user" in reply to "111"
-    Then ExoCom signals "users  --[ created-user ]->  web"
+    Then ExoCom signals "users  --[ created-user ]->  web  response time: 123456789ns"
```
Comment (kevgo) [inline features/broadcasting-messages.feature]: Let's just print:

```
users  --[ created-user ]->  web   (12ms)

File: src/listener-subsystem/zmq-listener.ls [hunk 2]
```diff
@@ -59,8 +59,10 @@ class ZMQListener extends EventEmitter
     name = req.name
     payload = req.payload
     response-to = req.response-to
+    timestamp = req.timestamp
+    response-time = req.response-time
     id = req.id
-    {sender, name, response-to, payload, id}
+    {sender, name, payload, id, response-to, timestamp, response-time}
```
Comment (kevgo) [inline src/listener-subsystem/zmq-listener.ls]: There are a number of entries here now, let's keep them sorted alphabetically.

**Comment by kevgo** [comment]
Nicely done! Let's see how this looks like with the shorter formatting.

**Comment by martinjaime** [comment]
@kevgo All comments addressed. 
image

File: features/support/api-world.ls [hunk 3]
```diff
@@ -70,16 +70,17 @@ ApiWorld = !->
...
   @verify-exocom-signaled-string = (string, done) ->
-    [sender-name, message-names, listeners] = string.split '  '
+    [sender-name, message-names, listeners, response-time-msg] = string.split '  '
     message-name-parts = message-names.split ' '
     switch message-name-parts.length
       | 3  =>  [_, translated-name, _] = message-name-parts
       | 5  =>  [_, original-name, _, translated-name, _] = message-name-parts
       | _  =>  throw new Error "Unknown message name parts"
     wait-until (~> @last-sent-messages[0].name is translated-name), 1, ~>
+      [seconds, nanoseconds] = @last-sent-messages[0].timestamp
```
Comment (kevgo) [inline features/support/api-world.ls]: Let's make the `timestamp` field a single number, consisting nanoseconds. This is hopefully a bit shorter and more compatible with other languages.

File: features/support/cli-world.ls [hunk 4]
```diff
@@ -81,7 +81,9 @@ CliWorld = !->
 
 
   @verify-exocom-signaled-string = (message, done) ->
-    @process.wait message, done
+    [...message-main, response-time-msg] = message.split '  '
+    response-time = response-time-msg.match(/-?\d+\.\d+/)?[0]
```
Comment (kevgo) [inline features/support/cli-world.ls]: You aren't using the local variable `response-time` here. Can you remove this line?

File: features/support/cli-world.ls [hunk 5]
```diff
@@ -81,7 +81,9 @@ CliWorld = !->
...
+    @process.wait "#{message-main.join '  '}  ", done
```
Comment (kevgo) [inline features/support/cli-world.ls]: I think overall this is a pretty brittle way of parsing this string. Let's try using a regular expression. How do you feel about something like this:

```ls
expected-string = message.match/(.*)  \(\d\.\d ms\)/[1]
@process.wait expected-string, done

**Comment by kevgo** [comment]
I love how the colored output looks like right now!

File: features/translating-messages.feature [hunk 6]
```diff
@@ -27,5 +27,5 @@ Feature: Translating messages
 
   Scenario: translating a reply
     When the "tweets" service sends "text-snippets.created" in reply to "111"
-    Then ExoCom signals "tweets  --[ text-snippets.created ]-[ tweets.created ]->  web"
+    Then ExoCom signals "tweets  --[ text-snippets.created ]-[ tweets.created ]->  web  response time: 123456789ns"
```
Comment (kevgo) [inline features/translating-messages.feature]: This is still the old format

File: src/cli.ls [hunk 7]
```diff
@@ -49,10 +49,14 @@ run = ->
 
     ..on 'message', ({messages, receivers}) ->
       for message in messages
+        response-time = ''
+        if message.response-to
+          [seconds, nanoseconds] = process.hrtime(message.timestamp)
+          response-time = "  (#{(seconds * 1e3 + nanoseconds / 1e6).to-fixed 1} ms)"
```
Comment (kevgo) [inline src/cli.ls]: This will become simpler when you switch to representing time as a single number of nanoseconds since the epoch.

Comment (kevgo) [inline src/cli.ls]: No need to store a separate string here. Just add it to line 57

File: src/listener-subsystem/http-listener.ls [hunk 8]
```diff
@@ -50,8 +50,8 @@ class HttpListener extends EventEmitter
       | _          =>  throw new Error "unknown error: #{result}"
 
 
-  _log: ({name, id, response-to}) ->
-    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}'"
+  _log: ({name, id, response-to, timestamp}) ->
+    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}' | response time: #{timestamp}"
```
Comment (kevgo) [inline src/listener-subsystem/http-listener.ls]: still old format here

**Comment by martinjaime** [comment]
@kevgo, I think this should be the last review. I could not find a module to give accurate nanosecond readings. `Date.now!` will give us the Unix timestamp we want with millisecond precision. We could also use `process.hrtime!` for nanosecond precision, but that will not give us a Unix timestamp. This means we would only be able to use the timestamp to measure response time.

**Comment by kevgo** [comment]
Please use `process.hrtime`. When you give it the original timestamp as an argument, it gives you the diff, see https://nodejs.org/api/process.html#process_process_hrtime_time. That should be all you need here.

File: features/translating-messages.feature [hunk 9]
```diff
@@ -26,6 +26,7 @@ Feature: Translating messages
 
 
   Scenario: translating a reply
-    When the "tweets" service sends "text-snippets.created" in reply to "111"
-    Then ExoCom signals "tweets  --[ text-snippets.created ]-[ tweets.created ]->  web"
+    When the "web" service sends "tweets.create"
+    And the "tweets" service sends "text-snippets.created" in reply to "111"
+    Then ExoCom signals "tweets  --[ text-snippets.created ]-[ tweets.created ]->  web  (8 ms)"
```
Comment (kevgo) [inline features/translating-messages.feature]: Are you sure it's always 8 ms here?

File: src/exocom.ls [hunk 10]
```diff
@@ -47,13 +47,12 @@ class ExoCom extends EventEmitter
...
   # sends the given message to all subscribers of it.
   send-message: (message-data) ~>
-    [seconds, nanos] = process.hrtime!
     # convert the outgoing message name from its internal version to the public version
     sender = @client-registry.clients[message-data.sender]
     external-message-name = @client-registry.outgoing-message-name message-data.name, sender
     message-data.original-name = message-data.name
     message-data.name = external-message-name
-    message-data.timestamp = seconds * 1e9 + nanos
+    message-data.timestamp = Date.now!
```
Comment (kevgo) [inline src/exocom.ls]: Please keep this as it was before. Before it was an integer timestamp with nanosecond precision. Which is what we want. Now it is a Date object with millisecond precision. We don't want that.

File: src/listener-subsystem/zmq-listener.ls [hunk 11]
```diff
@@ -59,8 +59,10 @@ class ZMQListener extends EventEmitter
     name = req.name
     payload = req.payload
     response-to = req.response-to
+    timestamp = req.timestamp
+    response-time = req.response-time
     id = req.id
-    {sender, name, response-to, payload, id}
+    {id, name, payload, response-time, response-to, sender, timestamp}
```
Comment (kevgo) [inline src/listener-subsystem/zmq-listener.ls]: Thanks for sorting this alphabetically!

File: src/listener-subsystem/zmq-listener.ls [hunk 12]
```diff
@@ -59,8 +59,10 @@ class ZMQListener extends EventEmitter
     name = req.name
     payload = req.payload
     response-to = req.response-to
+    timestamp = req.timestamp
+    response-time = req.response-time
```
Comment (kevgo) [inline src/listener-subsystem/zmq-listener.ls]: The incoming message should not contain the response time. ExoCom should calculate this message on its own.

File: src/listener-subsystem/http-listener.ls [hunk 13]
```diff
@@ -50,8 +50,8 @@ class HttpListener extends EventEmitter
       | _          =>  throw new Error "unknown error: #{result}"
 
 
-  _log: ({name, id, response-to}) ->
-    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}'"
+  _log: ({name, id, response-to, timestamp}) ->
+    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}'  (#{timestamp} ms)"
```
Comment (kevgo) [inline src/listener-subsystem/http-listener.ls]: I don't think we need to add the timing to the debug messages. The debug statement is rendered with its own timestamp.

File: src/listener-subsystem/zmq-listener.ls [hunk 14]
```diff
@@ -48,8 +48,8 @@ class ZMQListener extends EventEmitter
       | _                     =>  @emit 'error', "unknown result code: '#{@result}'"
 
 
-  _log: ({name, id, response-to}) ->
-    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}'"
+  _log: ({name, id, response-to, timestamp}) ->
+    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}'  (#{timestamp} ms)"
```
Comment (kevgo) [inline src/listener-subsystem/zmq-listener.ls]: Same here, this shouldn't be necessary.

File: src/cli.ls [hunk 15]
```diff
@@ -49,10 +49,13 @@ run = ->
 
     ..on 'message', ({messages, receivers}) ->
       for message in messages
+        response-time = ''
+        if message.response-to
+          response-time = "  (#{message.response-time} ms)"
```
Comment (kevgo) [inline src/cli.ls]: Make sure this is rounded properly. It should have 2 digits after the comma, i.e. look like "1.23 ms"

File: src/exocom.ls (hunk 10 above)
Comment (martinjaime) [inline src/exocom.ls]: Yes, I originally wanted to use `process.hrtime`, but switched when I found out that it was not Unix time. I will switch it, so we can have nanosecond precision.

File: src/message-sender/message-sender.ls [hunk 16]
```diff
@@ -44,13 +44,15 @@ class MessageSender extends EventEmitter
       timestamp: message-data.timestamp
     if message-data.response-to
       original-timestamp = @message-cache.get-original-timestamp request-data.id
+      request-data.response-time = Date.now! - original-timestamp
```
Comment (martinjaime) [inline src/message-sender/message-sender.ls]: @kevgo, the response time is calculated here.

**Comment by martinjaime** [comment]
@kevgo, I've moved all the response time logic from `message-sender` to the main `exocom` script. It turned out to be much more simpler. 

Also, added the `nanoseconds` npm module to do the simple conversion from the `process.hrtime` object into nanoseconds. It's simpler if we take the difference ourselves inline rather than calling the `process` API.

File: features/translating-messages.feature (hunk 9 above)
Comment (kevgo) [inline features/translating-messages.feature]: Can you say `(XX ms)` here to make clear that this is a placeholder?

File: package.json [hunk 17]
```diff
@@ -8,6 +8,7 @@
     "debug": "2.2.0",
     "docopt": "0.6.2",
     "express": "4.14.0",
+    "nanoseconds": "^0.1.0",
```
Comment (kevgo) [inline package.json]: Remove the `^`. Also, run `npm config set save-exact true` so that NPM does this correct from now on.

File: src/exocom.ls [hunk 18]
```diff
@@ -2,7 +2,9 @@ require! {
   './client-registry/client-registry' : ClientRegistry
   'events' : {EventEmitter}
   './listener-subsystem/listener-subsystem' : ListenerSubsystem
+  './message-cache/message-cache' : MessageCache
```
Comment (kevgo) [inline src/exocom.ls]: Can you sync this branch? I think these changes are unrelated.

File: src/exocom.ls [hunk 19]
```diff
@@ -2,7 +2,9 @@ require! {
...
   './message-sender/message-sender' : MessageSender
+  'nanoseconds' : Nanos
```
Comment (kevgo) [inline src/exocom.ls]: The convention is that identifiers with uppercase names are classes that need to be instantiated. Here you get a function back. I recommend naming it `nanoseconds`, similar to the module name.

File: src/exocom.ls [hunk 20]
```diff
@@ -47,18 +50,23 @@ class ExoCom extends EventEmitter
...
   send-message: (message-data) ~>
-    [seconds, nanos] = process.hrtime!
     # convert the outgoing message name from its internal version to the public version
     sender = @client-registry.clients[message-data.sender]
     external-message-name = @client-registry.outgoing-message-name message-data.name, sender
     message-data.original-name = message-data.name
     message-data.name = external-message-name
-    message-data.timestamp = seconds * 1e9 + nanos
-
+    message-data.timestamp = Nanos process.hrtime!
```
Comment (kevgo) [inline src/exocom.ls]: Great idea to use the nanoseconds library here!

File: src/message-cache/message-cache.ls [hunk 21]
```diff
@@ -1,5 +1,6 @@
 require! {
   'wait' : {repeat}
+  'nanoseconds' : Nanos
```
Comment (kevgo) [inline src/message-cache/message-cache.ls]: same here, don't name this like a class

File: src/message-cache/message-cache.ls [hunk 22]
```diff
@@ -9,8 +10,8 @@ class MessageCache
     @cleanup-interval = cleanup-interval
     @cache = {}
     repeat @cleanup-interval, ->
-      now = Date.now!
-      for id, timestamp of @cache when (timestamp - now) >= @cleanup-interval
+      now = Nanos process.hrtime!
```
Comment (kevgo) [inline src/message-cache/message-cache.ls]: No need to use nanosecond resolution here. You can keep it the old way. We just need to do regular cleanups at some point in time.  :)

File: src/exocom.ls (hunk 18 above)
Comment (martinjaime) [inline src/exocom.ls]: I had to move `message-cache` from `message-sender` to the main exocom script. So that is why this change is here. `message-cache` is unnecessary in `message-sender`. Should I create a new PR for that?

**Comment by martinjaime** [comment]
@kevgo, I addressed and replied to your comments

File: src/cli.ls [hunk 23]
```diff
@@ -49,10 +49,14 @@ run = ->
 
     ..on 'message', ({messages, receivers}) ->
       for message in messages
+        response-time = ''
+        if message.response-to
+          response-time-ms = message.response-time * 1e-6
```
Comment (kevgo) [inline src/cli.ls]: Inline this variable, it is used only once

File: src/exocom.ls (hunk 18 above)
Comment (kevgo) [inline src/exocom.ls]: Ah, very cool! No, this is perfect.

File: src/message-cache/message-cache.ls (hunk 22 above)
Comment (kevgo) [inline src/message-cache/message-cache.ls]: Ah, you compare the timestamp in the messages to the current time stamp. Makes sense!
